File: contracts/src/merkle.rs
use rustc_hash::FxHashMap;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

#[derive(Clone, Serialize, Deserialize)]
pub struct MerkleNode<T> {
    pub hash: String,
    pub value: T,
    pub children: FxHashMap<char, usize>,
    pub parent: usize,
}

fn hash_node<A: MerkleContainer<Vec<u8>> + Clone>(tree: &mut A, node_index: usize) -> String {
    let mut hasher = Sha256::new();
    hasher.update(tree.get_wrapper(node_index).unwrap().value.clone());
    for (_, index) in tree.get_wrapper(node_index).unwrap().children.iter() {
        hasher.update(tree.get_wrapper(index.clone()).unwrap().hash.clone());
    }
    let hash = hasher.finalize();
    hex::encode(hash)
}

fn fill_vec_node_hash(tree: &mut Vec<MerkleNode<Vec<u8>>>, node_index: usize) {
    tree[node_index].hash = hash_node(&mut tree.clone(), node_index);
}

fn hash_vec_tree(tree: &mut Vec<MerkleNode<Vec<u8>>>, index: usize) {
    if tree.len() == 1 {
        tree[0].hash = "".to_owned();
        return
    }
    if tree[index].children.len() == 0 {
        // Base case
        fill_vec_node_hash(tree, index);
    } else {
        // Recursive case
        for (_, child_index) in tree[index].children.clone().iter() {
            hash_vec_tree(tree, *child_index);
            fill_vec_node_hash(tree, index);
        }
    }
}

fn verify_node<A: MerkleContainer<Vec<u8>> + Clone>(tree: &mut A, node_index: usize) {
    assert_eq!(
        hash_node(tree, node_index),
        tree.get_wrapper(node_index).unwrap().hash,
        "Assertion failed on node with index {}",
        node_index
    );
    if node_index != 0 {
        verify_node(tree, tree.clone().get_wrapper(node_index).unwrap().parent);
    }
}

pub fn merklize(state: FxHashMap<String, Vec<u8>>) -> Vec<MerkleNode<Vec<u8>>> {
    let mut tree: Vec<MerkleNode<Vec<u8>>> = vec![MerkleNode {
        hash: String::new(),
        value: Vec::new(),
        children: FxHashMap::default(),
        parent: 0,
    }];
    for (loc, val) in state.iter() {
        let mut active: usize = 0;
        for c in loc.chars() {
            if tree[active].children.contains_key(&c) {
                active = tree[active].children[&c];
                continue;
            }
            tree.push(MerkleNode {
                hash: String::new(),
                value: Vec::new(),
                children: FxHashMap::default(),
                parent: active,
            });
            let new_index = tree.len() - 1;
            tree[active].children.insert(c, new_index);
            active = new_index;
        }
        tree[active].value = val.clone();
    }
    hash_vec_tree(&mut tree, 0);
    tree
}

pub trait MerkleContainer<T> {
    fn get_wrapper(&mut self, index: usize) -> Option<MerkleNode<T>>;
}

impl MerkleContainer<Vec<u8>> for Vec<MerkleNode<Vec<u8>>> {
    fn get_wrapper(&mut self, index: usize) -> Option<MerkleNode<Vec<u8>>> {
        self.get(index).cloned()
    }
}

pub fn get_from_merkle<A: MerkleContainer<Vec<u8>> + Clone>(tree: &mut A, loc: String) -> Vec<u8> {
    let mut active: usize = 0;
    for c in loc.chars() {
        active = tree.get_wrapper(active).unwrap().children[&c];
    }
    verify_node::<A>(tree, active);
    tree.get_wrapper(active).unwrap().value.clone()
}

File: node_util/merkle.go
package node_util

import (
	"crypto/sha256"
	"encoding/hex"
)

type MerkleNode struct {
	Data     []byte
	Children map[byte]int
	Parent   int
	Hash     string
	Key      string
}

func HashNode(tree []MerkleNode, nodeIndex int) string {
	hasher := sha256.New()
	hasher.Write(tree[nodeIndex].Data)
	for _, index := range tree[nodeIndex].Children {
		hasher.Write([]byte(tree[index].Hash))
	}
	hash := hasher.Sum(nil)
	return hex.EncodeToString(hash)
}

func HashTree(tree []MerkleNode, nodeIndex int) []MerkleNode {
	// Base case
	if len(tree[nodeIndex].Children) == 0 {
		tree[nodeIndex].Hash = HashNode(tree, nodeIndex)
	}
	// Recursive case
	for _, index := range tree[nodeIndex].Children {
		tree = HashTree(tree, index)
	}
	tree[nodeIndex].Hash = HashNode(tree, nodeIndex)
	return tree
}

func InsertValue(tree []MerkleNode, key string, value []byte) []MerkleNode {
	activeIndex := 0
	activeKey := ""
	if len(tree) == 0 {
		tree = append(tree, MerkleNode{
			Data:     make([]byte, 0),
			Children: make(map[byte]int),
			Parent:   0,
			Hash:     "",
			Key:      "",
		})
		activeIndex = 0
	}
	for i, _ := range key {
		activeKey += string(key[i])
		if _, ok := tree[activeIndex].Children[key[i]]; ok {
			activeIndex = tree[activeIndex].Children[key[i]]
			continue
		}
		tree[activeIndex].Children[key[i]] = len(tree)
		tree = append(tree, MerkleNode{
			Data:     make([]byte, 0),
			Children: make(map[byte]int),
			Parent:   activeIndex,
			Hash:     "",
			Key:      activeKey,
		})
		activeIndex = len(tree) - 1
	}
	tree[activeIndex].Data = []byte(value)
	tree = HashTree(tree, activeIndex)
	return tree
}

func GetValue(tree []MerkleNode, key string) ([]byte, bool) {
	activeIndex := 0
	for i, _ := range key {
		if val, ok := tree[activeIndex].Children[key[i]]; ok {
			activeIndex = val
		} else {
			return []byte(""), false
		}
	}
	return tree[activeIndex].Data, true
}

func Merge(initial []MerkleNode, delta []MerkleNode) []MerkleNode {
	// Get values from delta
	for _, node := range delta {
		if node.Data != nil {
			initial = InsertValue(initial, node.Key, node.Data)
		}
	}
	return initial
}
